<?xml version="1.0"?>
<launch>

  <!-- Tesseract Monitor Parameters -->
  <arg name="robot_description" default="robot_description"/>
  <arg name="discrete_plugin" default="BulletDiscreteBVHManager"/>
  <arg name="continuous_plugin" default="BulletCastBVHManager"/>
  <arg name="joint_state_topic" default=""/>
  <arg name="monitor_namespace" default="tesseract_workcell_environment"/>
  <arg name="cache_size" default="5"/>
  <arg name="cache_refresh_rate" default="0.1"/>
  <arg name="task_composer_config" default="$(find tesseract_task_composer)/config/task_composer_plugins.yaml"/>
  <arg name="input_key" default="input_data"/>
  <arg name="output_key" default="output_data"/>

  <!-- Launch Tesseract Monitor -->
  <node pkg="tesseract_monitoring" type="tesseract_monitoring_environment_node" name="tesseract_environment_monitor" output="screen" required="true">
    <param name="robot_description" type="string" value="$(arg robot_description)"/>
    <param name="discrete_plugin" type="string" value="$(arg discrete_plugin)"/>
    <param name="continuous_plugin" type="string" value="$(arg continuous_plugin)"/>
    <param name="joint_state_topic" type="string" value="$(arg joint_state_topic)"/>
    <param name="monitor_namespace" type="string" value="$(arg monitor_namespace)"/>
    <param name="publish_environment" type="bool" value="true"/>
  </node>

  <!-- Launch Tesseract Planning Server -->
  <!-- The planning server will monitor the environment namespace monitor_namespace for environment updates  -->
  <node pkg="twc_motion_planning" type="twc_motion_planning_planning_server_node" name="twc_planning_server" output="screen" required="true">
    <param name="robot_description" type="string" value="$(arg robot_description)"/>
    <param name="discrete_plugin" type="string" value="$(arg discrete_plugin)"/>
    <param name="continuous_plugin" type="string" value="$(arg continuous_plugin)"/>
    <param name="monitor_namespace" type="string" value="$(arg monitor_namespace)_planning_server"/>
    <param name="monitored_namespace" type="string" value="$(arg monitor_namespace)"/>
    <param name="cache_size" type="int" value="$(arg cache_size)"/>
    <param name="cache_refresh_rate" type="double" value="$(arg cache_refresh_rate)"/>
    <param name="task_composer_config" type="string" value="$(arg task_composer_config)"/>
    <param name="input_key" type="string" value="$(arg input_key)"/>
    <param name="output_key" type="string" value="$(arg output_key)"/>
  </node>

  <!-- Launch Tesseract Contact Monitor -->
  <!-- The planning server will monitor the environment namespace monitor_namespace for environment updates  -->
  <node name="tesseract_contact_monitoring" pkg="tesseract_monitoring" type="tesseract_monitoring_contacts_node" required="true">
    <param name="monitor_namespace" type="string" value="$(arg monitor_namespace)_contact_monitor"/>
    <param name="monitored_namespace" type="string" value="$(arg monitor_namespace)"/>
    <param name="publish_markers" value="true"/>
    <param name="publish_environment" value="true"/>
  </node>

</launch>
